### PR TITLE
Unit test plot_projection() from core/plotting/misc_plotting.py #413

### PR DESCRIPTION
**Pull request by Dylan-Celli**
@DanilYachmenev Here is the draft PR, please look at `core/plotting/notebooks/gallery_notebook` at the bottom for the plot that I made.
#413

### CODE REVIEW

**Comment by DanilYachmenev** [comment]
Ok I see

At first, pls also check in the `.py` copy of the notebook so I can comment on the code
Overall suggestion - try to create a bigger test dataframe with multiple columns 
For now it maybe due to the fact that you have only 1 column that is  being put on the axis

and also pls display your input `.head()` in the draft so it is also visible

**Comment by DanilYachmenev** [comment]
@Dylan-Celli any ETA? pls resolve conflicts and implement requested comments
no matter if the plot is not good but commit `.ipynb` and its `.py` copy

**Comment by Dylan-Celli** [comment]
> @Dylan-Celli any ETA? pls resolve conflicts and implement requested comments no matter if the plot is not good but commit `.ipynb` and its `.py` copy

I back on it today, sorry about last week I got busy moving out of my apartment. I expect to finish up with the test scenarios you [outlined](https://github.com/sorrentum/sorrentum/issues/413#issuecomment-1649614700) with in the next day.

File: core/plotting/notebooks/gallery_notebook.py [hunk 1]
```diff
@@ -66,3 +68,37 @@
...
+fig = plt.figure()
+ax = fig.add_axes([0,0,1,1])
+cpmiscplt.plot_projection(df, special_values = [1], mode = "scatter", ax = ax)
+
+# %%
+index = pd.date_range(start="2023-01-01", periods=len(data), freq="D")
+srs = pd.DataFrame(data, index=index)
+fig = plt.figure()
+ax = fig.add_axes([0,0,1,1])
+cpmiscplt.plot_projection(srs, special_values = [0], mode = "scatter", ax = ax)
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:100]: 2nd unit test should be different
read the end of https://github.com/sorrentum/sorrentum/issues/413#issuecomment-1649614700 for moe detailed desc

File: core/plotting/notebooks/gallery_notebook.py [hunk 2]
```diff
@@ -66,3 +68,37 @@
...
+    [1, 1, 0, 1],
+    [0, 1, 0, 1],
+    [0, 0, 1, 1],
+    [1, 1, 1, 1],
+    [1, 1, 1, 1],
+]
+df = pd.DataFrame(data)
+fig = plt.figure()
+ax = fig.add_axes([0,0,1,1])
+cpmiscplt.plot_projection(df, special_values = [0], ax = ax)
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:87]: run linter

File: core/plotting/notebooks/gallery_notebook.py [hunk 3]
```diff
@@ -66,3 +68,37 @@
...
+
+# %%
+# Calling plot_projection with the scatter plot mode enabled
+fig = plt.figure()
+ax = fig.add_axes([0,0,1,1])
+cpmiscplt.plot_projection(df, special_values = [1], mode = "scatter", ax = ax)
+
+# %%
+index = pd.date_range(start="2023-01-01", periods=len(data), freq="D")
+srs = pd.DataFrame(data, index=index)
```
[the remark this answers is not in the document]
Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py:97]: I just wanted to make the variable naming convention consistent

File: core/plotting/notebooks/gallery_notebook.py [hunk 4]
```diff
@@ -66,3 +68,37 @@
...
+fig = plt.figure()
```
[the remark this answers is not in the document]
Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py:98]: Ok, I can remove it. I thought that a new fig was necessary when doing multiple plots.

File: core/plotting/notebooks/gallery_notebook.py [hunk 5]
```diff
@@ -66,3 +68,37 @@
...
+ax = fig.add_axes([0,0,1,1])
```
[the remark this answers is not in the document]
Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py:99]: Ok, I will remove

File: core/plotting/notebooks/gallery_notebook.py [hunk 6]
```diff
@@ -66,3 +68,37 @@
...
+df = pd.DataFrame(data)
+fig = plt.figure()
+ax = fig.add_axes([0,0,1,1])
+cpmiscplt.plot_projection(df, special_values = [0], ax = ax)
+
+# %%
+# Calling plot_projection with the scatter plot mode enabled
+fig = plt.figure()
+ax = fig.add_axes([0,0,1,1])
+cpmiscplt.plot_projection(df, special_values = [1], mode = "scatter", ax = ax)
```
[the remark this answers is not in the document]
Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py:93]: Sorry, I misinterpreted you earlier message about leaving the output for the code. I thought you meant that each time I committed I should clear the output for any notebook I used. I will make sure to keep the output this time

File: core/plotting/notebooks/gallery_notebook.py (hunk 1 above)
Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py:100]: Sorry, I left the code I used to experiments with the function output to see what it looked like. I fixed it to align with your comment

**Comment by Dylan-Celli** [comment]
> also fix merge conflicts

Should I just take all the changes from master? I saw that most of the conflicts related to the cell's metadata changing but I also see some code changes so I want to double check before I resolve the conflict.

**Comment by DanilYachmenev** [comment]
> > also fix merge conflicts
> 
> @DanilYachmenev Should I just take all the changes from master? I saw that most of the conflicts related to the cell's metadata changing but I also see some code changes so I want to double check before I resolve the conflict.

merge master changes and adjust your code accordingly
it may change later on again but yeah, in general consider master changes as the major ones

File: core/plotting/notebooks/gallery_notebook.py [hunk 7]
```diff
@@ -150,4 +150,28 @@
...
+# Calling plot_projection to test that it's basic functionality works.
+data = [
+    [1, 1, 0, 1],
+    [0, 1, 0, 1],
+    [0, 0, 1, 1],
+    [1, 1, 1, 1],
+    [1, 1, 1, 1],
+]
+index = pd.date_range(start="2023-01-01", periods=len(data), freq="D")
+df = pd.DataFrame(data, index=index)
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:166]: use `get_plot_projection1()` when created

File: core/plotting/test/test_plots.py [hunk 8]
```diff
@@ -170,3 +170,20 @@ def test_plot_spectrum2(self) -> None:
...
+        """
+        data = [
+            [1, 1, 0, 1],
+            [0, 1, 0, 1],
+            [0, 0, 1, 1],
+            [1, 1, 1, 1],
+            [1, 1, 1, 1],
+        ]
+        index = pd.date_range(start="2023-01-01", periods=len(data), freq="D")
+        srs = pd.Series(data, index=index)
```
Comment (DanilYachmenev) [inline core/plotting/test/test_plots.py:186]: why is it `srs` btw?

File: core/plotting/notebooks/gallery_notebook.py [hunk 9]
```diff
@@ -150,4 +150,28 @@
...
+fig = plt.figure()
+ax = fig.add_axes([0, 0, 1, 1])
+cplmiplo.plot_projection(df, special_values=[0], ax=ax)
+
+# %%
+# Calling plot_projection with the scatter plot mode enabled.
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:172]: remove, self-explanatory call

File: core/plotting/notebooks/gallery_notebook.py [hunk 10]
```diff
@@ -150,4 +150,28 @@
...
+fig = plt.figure()
+ax = fig.add_axes([0, 0, 1, 1])
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:174]: lets not use it in 1 of the calls, redundant double-check that they work

File: core/plotting/notebooks/gallery_notebook.py [hunk 11]
```diff
@@ -150,4 +150,28 @@
...
+cplmiplo.plot_projection(df, special_values=[0], mode="scatter", ax=ax)
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:175]: set `mode` outside the func call

File: core/plotting/notebooks/gallery_notebook.py [hunk 12]
```diff
@@ -150,4 +150,28 @@
 axes_flat = axes.flatten()
 cplmiplo.plot_spectrum(signal=test_df, axes=axes_flat)
 
+# %% [markdown]
+# ## `plot_projection()`
+
+# %%
+# Calling plot_projection to test that it's basic functionality works.
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:157]: -> `# Set inputs.`

File: core/plotting/test/test_plots.py (hunk 8 above)
Comment (Dylan-Celli) [reply core/plotting/test/test_plots.py:186]: I changed it to `df`, I thought that it was more consistent with the current convention initially.

File: README.md [hunk 13]
```diff
@@ -113,8 +113,6 @@ This contains a set of resources that over time will make 10x more productive
 
 - [Design Philosophy](docs/Design_Philosophy.md)
 
-- [Codebase Clean-up](docs/Codebase_clean_up_scripts.md)
-
 # Technologies used
```
[the remark this answers is not in the document]
Comment (DanilYachmenev) [reply README.md:116]: not implemented
bring these lines back top the doc so this change disappears from here

File: README.md [hunk 14]
```diff
@@ -140,8 +138,6 @@ This contains a set of resources that over time will make 10x more productive
 
 - [Telegram](/docs/Telegram.md)
 
-- [Email](/docs/Email.md)
-
```
Comment (DanilYachmenev) [inline README.md:144]: this one as well

File: core/plotting/test/test_plots.py [hunk 15]
```diff
@@ -170,3 +170,20 @@ def test_plot_spectrum2(self) -> None:
         _, axes = plt.subplots(2, 2, figsize=(20, 10))
         axes_flat = axes.flatten()
         cplmiplo.plot_spectrum(signal=test_df, axes=axes_flat)
+
+    def test_plot_projection(self) -> None:
```
[the remark this answers is not in the document]
Comment (DanilYachmenev) [reply core/plotting/test/test_plots.py]: not implemented
now you have 2 equal tests -what for?
read about the test specs in the comment pls 
```
for the 2nd test replace all 0 in the test data to None and call the function without special_values, but use mode="scatter"
```

File: core/plotting/notebooks/gallery_notebook.py [hunk 16]
```diff
@@ -150,4 +150,21 @@
...
+
+# %%
+cplmiplo.plot_projection(df, special_values=[0])
+
+# %%
+# Calling plot_projection with the scatter plot mode enabled.
+fig = plt.figure()
+ax = fig.add_axes([0, 0, 1, 1])
+mode = "scatter"
+cplmiplo.plot_projection(df, special_values=[0], mode=mode#), ax=ax)
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:168]: and you need to make another call for the 2nd test

File: core/plotting/notebooks/gallery_notebook.py [hunk 17]
```diff
@@ -150,4 +150,21 @@
...
+
+# %%
+# Set inputs.
+df = cptetepl.Test_plots.get_plot_projection1()
+
+# %%
+cplmiplo.plot_projection(df, special_values=[0])
+
+# %%
+# Calling plot_projection with the scatter plot mode enabled.
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:164]: remove comments from the notebook code

File: core/plotting/notebooks/gallery_notebook.py [hunk 18]
```diff
@@ -150,4 +150,21 @@
 axes_flat = axes.flatten()
 cplmiplo.plot_spectrum(signal=test_df, axes=axes_flat)
 
+# %% [markdown]
+# ## `plot_projection()`
+
+# %%
+# Set inputs.
```
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:157]: this one too

File: core/plotting/notebooks/gallery_notebook.py (hunk 16 above)
Comment (DanilYachmenev) [inline core/plotting/notebooks/gallery_notebook.py:168]: what about `ax`?
you can use it in 1 example and do not use in another to demonstrate how it impacts plots

File: dev_scripts/encrypt_model.py [hunk 19]
```diff
@@ -1,195 +1,107 @@
 #!/usr/bin/env python
 """
```
Comment (DanilYachmenev) [inline dev_scripts/encrypt_model.py:2]: this file should not be here, no changes expected
same for all the other files

**Review by DanilYachmenev — COMMENTED** [review]
you need to commit changes only to files that are related to the issue
check yourself please at `Files changed` section 

if you cannot revert the changes then I recommend you to branch from the pulled master and file a new PR, just do not forget to close this one and remove the related branch

File: core/plotting/notebooks/gallery_notebook.py (hunk 16 above)
Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py]: I added the axis argument to test functionality, to my knowledge it should not impact the graphing of the plots.

Comment (Dylan-Celli) [reply core/plotting/notebooks/gallery_notebook.py]: I changed this second call to match the description that you provided, removing and replacing the 0s in the df with `None` values and calling the function without the `special_values` argument.

**Comment by Dylan-Celli** [comment]
@DanilYachmenev Would it be better to create a new branch and close this out if I keep getting merge conflicts on `core/plotting/notebooks/gallery_notebook.ipynb` and `core/plotting/notebooks/gallery_notebook.py`? Whenever I push my commit the metadata between the branches does not align and I keep having to fix it. If I use `git merge` with master it will create a commit that affects files outside of the three that I need, but if I just try to checkout my three files from master it creates a merge conflict after pushing. I think that my branch being out-of-date is the major problem.

**Comment by DanilYachmenev** [comment]
@Dylan-Celli its ok, sometimes merge conflicts are very annoying, especially with notebooks

just do not forget to reference this one in a new PR, delete this branch and close this PR

**Pull request closed without merging**